tests/raftstore/transport_simulate.rs

```diff
@@ -0,0 +1,103 @@
+use kvproto::raft_serverpb::RaftMessage;
+use tikv::raftstore::Result;
+use tikv::raftstore::store::Transport;
+use rand;
+use std::sync::{Arc, RwLock};
+
+use super::util::*;
+use self::Strategy::*;
+
+#[derive(Clone)]
+pub enum Strategy {
+    Default,
+    LossPacket(u32),
+    Latency(u64),
+    OutOrder,
+}
+
+trait Filter: Send + Sync {
+    fn before(&self, msg: &RaftMessage) -> bool;
+    fn after(&self, Result<()>) -> Result<()>;
+}
+
+struct FilterLossPacket(u32);
+struct FilterLatency(u64);
+struct FilterOutOrder;
+
+impl Filter for FilterLossPacket {
+    fn before(&self, _: &RaftMessage) -> bool {
+        rand::random::<u32>() % 100u32 < self.0
+    }
+    fn after(&self, x: Result<()>) -> Result<()> {
+        x
+    }
+}
+
+impl Filter for FilterLatency {
+    fn before(&self, _: &RaftMessage) -> bool {
+        sleep_ms(self.0);
+        false
+    }
+    fn after(&self, x: Result<()>) -> Result<()> {
+        x
+    }
+}
+
+impl Filter for FilterOutOrder {
+    fn before(&self, _: &RaftMessage) -> bool {
+        unimplemented!()
+    }
+    fn after(&self, _: Result<()>) -> Result<()> {
+        unimplemented!()
+    }
+}
+
+pub struct SimulateTransport<T: Transport> {
+    filters: Vec<Box<Filter>>,
+    trans: Arc<RwLock<T>>,
+}
+
+impl<T: Transport> SimulateTransport<T> {
+    pub fn new(strategy: Vec<Strategy>, trans: Arc<RwLock<T>>) -> SimulateTransport<T> {
+        let mut filters: Vec<Box<Filter>> = vec![];
+        for s in strategy {
+            match s {
+                Default => {
+                    continue;
+                }
+                LossPacket(rate) => {
+                    filters.push(Box::new(FilterLossPacket(rate)));
+                }
+                Latency(latency) => {
+                    filters.push(Box::new(FilterLatency(latency)));
+                }
+                OutOrder => {
+                    filters.push(Box::new(FilterOutOrder));
+                }
+            }
+        }
+
+        SimulateTransport {
+            filters: filters,
+            trans: trans,
+        }
+    }
+}
+
+impl<T: Transport> Transport for SimulateTransport<T> {
+    fn send(&self, msg: RaftMessage) -> Result<()> {
+        for strategy in &self.filters {
+            if strategy.before(&msg) {
+                return strategy.after(Ok(()));
+            }
+        }
+
+        let mut res = self.trans.read().unwrap().send(msg);
+
+        for strategy in &self.filters {
```

Comment:
I think it's needed. Consider that if I want to simulate some error, the packet send may be success or fail. With after handling provided we can hijack the result and return what error we want. 

take Timeout for example, we don't know whether send is success or not, we can simulate either situation and return Timeout error in after handling.

Later I shall write something like FilterErrorOccur.